src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs

```diff
@@ -10,20 +10,84 @@ namespace Aspire.Hosting.ApplicationModel;
+    /// </summary>
+    /// <param name="name">The name of the resource.</param>
+    /// <param name="userNameParameter">A parameter that contains the MongoDb server user name, or <see langword="null"/> to use a default value.</param>
+    /// <param name="passwordParameter">A parameter that contains the MongoDb server password.</param>
+    public MongoDBServerResource(string name, ParameterResource? userNameParameter, ParameterResource? passwordParameter) : this(name)
+    {
+        UserNameParameter = userNameParameter;
+        PasswordParameter = passwordParameter;
+    }
+
     /// <summary>
     /// Gets the primary endpoint for the MongoDB server.
     /// </summary>
     public EndpointReference PrimaryEndpoint => _primaryEndpoint ??= new(this, PrimaryEndpointName);
 
+    /// <summary>
+    /// Gets the parameter that contains the MongoDb server password.
+    /// </summary>
+    public ParameterResource? PasswordParameter { get; }
+        
+    /// <summary>
+    /// Gets the parameter that contains the MongoDb server username.
+    /// </summary>
+    public ParameterResource? UserNameParameter { get; }
+
+    internal ReferenceExpression UserNameReference =>
+        UserNameParameter is not null ?
+            ReferenceExpression.Create($"{UserNameParameter}") :
+            ReferenceExpression.Create($"{DefaultUserName}");
+
     /// <summary>
     /// Gets the connection string for the MongoDB server.
     /// </summary>
-    public ReferenceExpression ConnectionStringExpression =>
-        ReferenceExpression.Create(
-            $"mongodb://{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
+    public ReferenceExpression ConnectionStringExpression => ReferenceExpression.Create($"{ConnectionStringWithoutOptionsExpression}{ConnectionStringOptionsExpression}");
+
+    /// <summary>
+    /// Gets the connection string for the MongoDB server without options parameters.
+    /// </summary>
+    internal ReferenceExpression ConnectionStringWithoutOptionsExpression
+    {
+        get
+        {
+            if (PasswordParameter is null)
+            {
+                return ReferenceExpression.Create($"mongodb://{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
+            }
+            else
+            {
+                return ReferenceExpression.Create($"mongodb://{UserNameReference}:{PasswordParameter}@{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
+            }
+        }
+    }
+
+    /// <summary>
+    /// Gets the options parameters for connection string of the MongoDB server.
+    /// </summary>
+    internal ReferenceExpression ConnectionStringOptionsExpression
+    {
+        get
+        {
+            if (PasswordParameter is null)
+            {
+                return ReferenceExpression.Create($"");
```

Comment:
I don't love this pattern. We have a ReferenceExpressionBuilder that could clean up some of this I think. We could have a single method that took a builder and some arguments and returned an expression.

Reply:
Fixed.